### PR TITLE
TDD - Add query algorithm and unit tests with optional attendees

### PR DESCRIPTION
**Pull request by antoniolinhart**
- Add query algorithm, broken up into multiple methods
  - Accounts for optional attendees (either *all* or *none*)
- Passes all initial tests
- Passes additional tests

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -121,6 +123,68 @@ public void everyAttendeeIsConsidered() {
...
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+    @Test
```
Comment (VivekShankar) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:157]: nit: align @Test with the test method signature.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 2]
```diff
@@ -223,6 +287,34 @@ public void justEnoughRoom() {
     Assert.assertEquals(expected, actual);
   }
 
+  @Test
+  public void justEnoughRoomDoNotConsiderOptionalAttendee() {
```
Comment (VivekShankar) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:291]: Nice - this is a good test case to include!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,170 @@
...
+  /**
+   * Determine available times for a meeting to take place given a list of Events and a Meeting Request.
+   * @param events the list of events to take into consideration when finding available times
+   * @param request the meeting request with meeting attendees and duration
+   * @return A list of available TimeRanges
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Meeting cannot last longer than 24 hours
+    if(request.getDuration() > TimeRange.WHOLE_DAY.duration()) return Arrays.asList();
```
Comment (VivekShankar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:34]: nit: Here and below, can you add a "space" between the `if` and the initial left paren "(" of the condition?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,170 @@
...
+    // If there are no available times with all (mandatory and optional attendees) then only consider mandatory 
+    if(availableTimesWithOptionalAttendees.size() == 0 && request.getAttendees().size() > 0) {
+      return queryHelper(events, request, false);
+    }
+
+    return availableTimesWithOptionalAttendees;
+  }
+
+  /**
+   * Helper method for the query method which finds the available times to meet given a list of events and a MeetingRequest,
```
Comment (VivekShankar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:47]: The Java style guide suggests a column limit of 100 characters (https://google.github.io/styleguide/javaguide.html#s4.4-column-limit). Can you wrap this line and those below to stick to the 100 character line limit?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,170 @@
...
+    Collection<String> attendees = new HashSet<String>(request.getAttendees());
+    if(considerOptional) {
+      attendees.addAll(request.getOptionalAttendees());
+    }
+    
+    // Needs to have at least 1 attendee at a meeting
+    if(attendees.size() == 0) return Arrays.asList(TimeRange.WHOLE_DAY);
+
+    // For a time to be unavailable, there must be someone at the event who is in the meeting attendees list
+    List<TimeRange> unavailableTimes = new ArrayList(convertAttendeesAndEventsToUnavailableTimes(events, attendees));
```
Comment (VivekShankar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:65]: I really like the way you've structured your `queryHelper` function - having helpers to do each of the sub-tasks here makes for clear, readable code that's easy to follow!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,170 @@
...
+  }
+
+  /**
+   * Removes items from a Collection if they are shorter than the specified duration.
+   * @param times a collection of TimeRanges
+   * @param duration the threshold length of the meeting 
+   * @return a new collection with a list of sufficient times at least as long as the given duration
+   */
+  private static Collection<TimeRange> removeAllTimeRangesLessThanGivenDuration(Collection<TimeRange> times, long duration) {
+    Collection<TimeRange> sufficientTimes = new ArrayList<>();
```
Comment (VivekShankar) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:88]: I'm not sure exactly which version of Java you're using for the walkthroughs, but have you used Java Collections and Streams before (they're available in Java 8 and above)? I think this method and a couple below could be good opportunities to try them out (here's a sample tutorial with a few examples: http://tutorials.jenkov.com/java-collections/streams.html).

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 7]
```diff
@@ -121,6 +123,68 @@ public void everyAttendeeIsConsidered() {
     Assert.assertEquals(expected, actual);
   }
 
+  @Test
+  public void everyAttendeeIsConsideredWithOptionalAttendeeBlockingEntireDay() {
```
Comment (VivekShankar) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:127]: Just a fyi: for consistency, we follow the "arrange, act, test" model of writing Java unit tests at Google!

The **arrange** step is for setting up details of the external world specific to the situation under test (e.g. in this case, initializing `events`, `request`).

The **act** step is where you actually make calls to the class/object under test. In this case: `query.query(...)`.

The **assert** step is the place to make assertions on the return values collected.

We usually leave a blank line in between the arrange, act, and test steps.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 8]
```diff
@@ -223,6 +287,34 @@ public void justEnoughRoom() {
...
+  public void justEnoughRoomDoNotConsiderOptionalAttendee() {
+    // Have one mandatory person and one optional person, ignore the optional person since it would make
+    // the meeting too short.
+    //
+    // Events  : |--A--|     |----A----|
+    // Optional:       |B|
+    // Day     : |---------------------|
+    // Options :       |-----|
+
+    System.out.println("\n\nfail test");
```
Comment (antoniolinhart) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:300]: Didn't remove test println!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (antoniolinhart) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks! There's some logic in places that doesn't make clear sense without comments so I wanted to break it down

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (antoniolinhart) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Didn't know about Streams! Thank you!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 7 above)
Comment (antoniolinhart) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:127]: Fixed!

**Pull request merged**